pkg/kv/cosmosdb/store.go

```diff
@@ -0,0 +1,434 @@
+		}
+		return nil, err
+	}
+
+	var itemResponseBody Document
+	err = json.Unmarshal(itemResponse.Value, &itemResponseBody)
+	if err != nil {
+		return nil, err
+	}
+
+	val, err := encoding.DecodeString(itemResponseBody.Value)
+	if err != nil {
+		return nil, err
+	}
+	return &kv.ValueWithPredicate{
+		Value:     val,
+		Predicate: kv.Predicate([]byte(itemResponse.ETag)),
+	}, nil
+}
+
+func errStatusCode(err error) int {
+	var respErr *azcore.ResponseError
+	if !errors.As(err, &respErr) {
+		return -1
+	}
+	return respErr.StatusCode
+}
+func isErrStatusCode(err error, code int) bool {
+	return errStatusCode(err) == code
+}
+
+func (s *Store) Set(ctx context.Context, partitionKey, key, value []byte) error {
+	if len(partitionKey) == 0 {
+		return kv.ErrMissingPartitionKey
+	}
+	if len(key) == 0 {
+		return kv.ErrMissingKey
+	}
+	if value == nil {
+		return kv.ErrMissingValue
+	}
+
+	// Specifies the value of the partiton key
+	item := Document{
+		PartitionKey: encoding.EncodeToString(partitionKey),
+		HashID:       s.hashID(key),
+		SortID:       encoding.EncodeToString(key),
+		Value:        encoding.EncodeToString(value),
+	}
+
+	b, err := json.Marshal(item)
+	if err != nil {
+		return err
+	}
+	itemOptions := azcosmos.ItemOptions{
+		ConsistencyLevel: s.consistencyLevel.ToPtr(),
+	}
+	pk := azcosmos.NewPartitionKeyString(item.PartitionKey)
+
+	_, err = s.containerClient.UpsertItem(ctx, pk, b, &itemOptions)
+	return err
+}
+
+func (s *Store) SetIf(ctx context.Context, partitionKey, key, value []byte, valuePredicate kv.Predicate) error {
+	if len(partitionKey) == 0 {
+		return kv.ErrMissingPartitionKey
+	}
+	if len(key) == 0 {
+		return kv.ErrMissingKey
+	}
+	if value == nil {
+		return kv.ErrMissingValue
+	}
+
+	// Specifies the value of the partiton key
+	item := Document{
+		PartitionKey: encoding.EncodeToString(partitionKey),
+		HashID:       s.hashID(key),
+		SortID:       encoding.EncodeToString(key),
+		Value:        encoding.EncodeToString(value),
+	}
+
+	b, err := json.Marshal(item)
+	if err != nil {
+		return err
+	}
+	itemOptions := azcosmos.ItemOptions{
+		ConsistencyLevel: s.consistencyLevel.ToPtr(),
+	}
+	pk := azcosmos.NewPartitionKeyString(item.PartitionKey)
+
+	switch valuePredicate {
+	case nil:
+		_, err = s.containerClient.CreateItem(ctx, pk, b, &itemOptions)
+		if isErrStatusCode(err, http.StatusConflict) {
+			return kv.ErrPredicateFailed
+		}
+	case kv.PrecondConditionalExists:
+		patch := azcosmos.PatchOperations{}
+		patch.AppendReplace("/value", item.Value)
+		_, err = s.containerClient.PatchItem(
+			ctx,
+			pk,
+			item.HashID,
+			patch,
+			&itemOptions,
+		)
+		if isErrStatusCode(err, http.StatusNotFound) {
+			return kv.ErrPredicateFailed
+		}
+	default:
+		etag := azcore.ETag(valuePredicate.([]byte))
+		itemOptions.IfMatchEtag = &etag
+		_, err = s.containerClient.UpsertItem(ctx, pk, b, &itemOptions)
+		if isErrStatusCode(err, http.StatusPreconditionFailed) {
+			return kv.ErrPredicateFailed
+		}
```

Comment:
do we need to also handle key not found?

Reply:
With Upsert? impossible to have a key not found.